File: mobile/app/src/main/java/com/hebit/app/domain/model/Habit.kt
package com.hebit.app.domain.model

import java.time.LocalDate
import java.time.LocalDateTime

// Using java.time for dates and times. Ensure your project's minSdk supports this
// or has core library desugaring enabled.

enum class HabitFrequency {
    DAILY, WEEKLY, MONTHLY, SPECIFIC_DATES, UNKNOWN
}

enum class HabitStatus {
    ACTIVE, ARCHIVED, UNKNOWN
}

enum class HabitDifficulty {
    EASY, MEDIUM, HARD, UNKNOWN
}

enum class ReminderNotificationStyle {
    BASIC, MOTIVATIONAL, UNKNOWN
}

enum class SuccessCriteriaType {
    BOOLEAN, NUMERIC, TIMER, UNKNOWN
}

data class HabitFrequencyConfig(
    val daysOfWeek: List<Int>? = null, // 0 (Sun) - 6 (Sat)
    val datesOfMonth: List<Int>? = null, // 1-31, or -1 for last day
    val timesPerPeriod: Int? = null,
    val specificDates: List<LocalDate>? = null
)

data class HabitStreakData(
    val current: Int?,
    val longest: Int?,
    val lastCompleted: LocalDateTime? = null
)

data class HabitCompletionHistoryEntry(
    val date: LocalDateTime,
    val completed: Boolean,
    val notes: String? = null,
    val skipReason: String? = null,
    val value: Int? = null, // For measurable habits
    val mood: Int? = null // 1-5 mood rating
)

data class ReminderSettings(
    val time: String? = null, // Consider LocalTime if appropriate
    val customMessage: String? = null
)

data class SuccessCriteria(
    val type: SuccessCriteriaType?,
    val target: Int? = null
)

data class HabitMetadata(
    val successRate: Float? = null
)

data class Habit(
    val id: String,
    val userId: String,
    val title: String,
    val description: String? = null,
    val icon: String? = null,
    val color: String? = null,
    val frequency: HabitFrequency,
    val frequencyConfig: HabitFrequencyConfig? = null,
    val streakData: HabitStreakData? = null,
    val category: String? = null, // Could be a Category domain model if it exists
    val completionHistory: List<HabitCompletionHistoryEntry> = emptyList(),
    val status: HabitStatus = HabitStatus.ACTIVE,
    val difficulty: HabitDifficulty? = null,
    val impact: Int? = null,
    val startDate: LocalDate? = null,
    val endDate: LocalDate? = null,
    val reminderSettings: ReminderSettings? = null,
    val successCriteria: SuccessCriteria? = null,
    val metadata: HabitMetadata? = null,
    val createdAt: LocalDateTime? = null,
    val updatedAt: LocalDateTime? = null,
    val completedToday: Boolean? = null // Derived information
)

File: backend/src/models/Habit.ts
import mongoose, { Schema } from 'mongoose';
import { HabitDocument } from '../types';

const habitSchema = new Schema<HabitDocument>(
  {
    user: {
      type: String,
      required: true,
      ref: 'User'
    },
    title: {
      type: String,
      required: [true, 'Please add a title'],
      trim: true
    },
    description: {
      type: String,
      trim: true
    },
    frequency: {
      type: String,
      enum: ['daily', 'weekly', 'monthly'],
      default: 'daily'
    },
    frequencyConfig: {
      daysOfWeek: [
        {
          type: Number,
          min: 0,
          max: 6
        }
      ],
      datesOfMonth: [
        {
          type: Number,
          min: 1,
          max: 31
        }
      ],
      timesPerPeriod: {
        type: Number,
        default: 1,
        min: 1
      }
    },
    timeOfDay: {
      type: String
    },
    daysOfWeek: [
      {
        type: Number,
        min: 0,
        max: 6
      }
    ],
    timePreference: {
      preferredTime: {
        type: String
      },
      flexibility: {
        type: Number,
        default: 30, // 30 minutes flexibility
        min: 0
      }
    },
    streak: {
      type: Number,
      default: 0
    },
    streakData: {
      current: {
        type: Number,
        default: 0
      },
      longest: {
        type: Number,
        default: 0
      },
      lastCompleted: {
        type: Date
      }
    },
    category: {
      type: String,
      ref: 'Category'
    },
    completionHistory: [
      {
        date: {
          type: Date,
          required: true
        },
        completed: {
          type: Boolean,
          required: true
        },
        value: {
          type: Number
        },
        notes: {
          type: String
        },
        mood: {
          type: Number,
          min: 1,
          max: 5
        },
        skipReason: {
          type: String
        }
      }
    ],
    difficulty: {
      type: String,
      enum: ['easy', 'medium', 'hard'],
      default: 'medium'
    },
    startDate: {
      type: Date,
      default: Date.now
    },
    endDate: {
      type: Date
    },
    reminderSettings: {
      time: {
        type: String
      },
      customMessage: {
        type: String
      },
      notificationStyle: {
        type: String,
        enum: ['basic', 'motivational'],
        default: 'basic'
      }
    },
    successCriteria: {
      type: {
        type: String,
        enum: ['boolean', 'numeric', 'timer'],
        default: 'boolean'
      },
      target: {
        type: Number
      },
      unit: {
        type: String
      },
      minimumThreshold: {
        type: Number
      }
    },
    metadata: {
      successRate: {
        type: Number,
        min: 0,
        max: 100,
        default: 0
      },
      averageCompletionTime: {
        type: String
      },
      contextPatterns: {
        location: [String],
        precedingActivities: [String],
        followingActivities: [String]
      }
    }
  },
  {
    timestamps: true
  }
);

// Index for faster queries
habitSchema.index({ user: 1 });
habitSchema.index({ frequency: 1 });
habitSchema.index({ 'streakData.current': -1 });
habitSchema.index({ difficulty: 1 });
habitSchema.index({ 'completionHistory.date': 1 });

const Habit = mongoose.model<HabitDocument>('Habit', habitSchema);

export default Habit;
